src/ConfirmProvider.js

```diff
@@ -40,36 +40,56 @@ const buildOptions = (defaultOptions, options) => {
 const ConfirmProvider = ({ children, defaultOptions = {} }) => {
   const [options, setOptions] = useState({ ...DEFAULT_OPTIONS, ...defaultOptions });
   const [resolveReject, setResolveReject] = useState([]);
-  const [resolve, reject] = resolveReject;
+  const [resolve, reject] = resolveReject
+  /**
+   * Possible states:
+   * 1. "Idle" initially
+   * 2. "Open" when `confirm` is called
+   * 3. "Submitting" when confimation or cancellation button is pressed.
+   * After submitting, the state is back to "Idle" since the Dialog is closed.
+   */
+  const [state, setState] = useState('idle');
 
   const confirm = useCallback((options = {}) => {
     return new Promise((resolve, reject) => {
       setOptions(buildOptions(defaultOptions, options));
       setResolveReject([resolve, reject]);
+      setState('open')
     });
   }, []);
 
   const handleClose = useCallback(() => {
     setResolveReject([]);
+    setState('idle');
   }, []);
 
   const handleCancel = useCallback(() => {
-    reject();
-    handleClose();
-  }, [reject, handleClose]);
+    if (state === 'open') {
+      reject();
+      setState('submitting');
+    }
+  }, [reject]);
 
   const handleConfirm = useCallback(() => {
-    resolve();
-    handleClose();
-  }, [resolve, handleClose]);
+    if (state === 'open') {
+      resolve();
+      setState('submitting');
+    }
+  }, [resolve, state]);
+
+  useEffect(() => {
+    if (state === 'submitting') {
+      handleClose();
```

Comment:
`handleClose` is always executed after confirm/cancel, because we pass it as `onClose` to the dialog, so I think this would call it twice.

Reply:
Correct, that is why I updated I moved "calling `handleClose`" inside the `useEffect` to make this "transactional".

**Transactions being:**
1. Click on confirm => call `handleConfirm` => set state to "submitting" => Inside `useEffect`, call `handleClose`
2. Click on cancel => call `handleCancel` => set state to "submitting" => Inside `useEffect`, call `handleClose`

If a user clicks the confirm button twice then, before the second click, the state is already in `submitting` state, so it won't submit the form again solving the issue at hand

Reply:
Yeah, my point is that we still pass `onClose={handleClose}` to the dialog, so it will be called by the dialog too, so currently it is:

1. Click on confirm => call `handleConfirm` => set state to "submitting" => call `handleClose` => Inside `useEffect`, call `handleClose`

Reply:
> Click on confirm => call handleConfirm => set state to "submitting" => call handleClose => Inside useEffect, call handleClose

The updated code of `handleConfirm` in this PR, won't call `handleClose`.

In the new code, handleClose is called from two places only:
1. If someone clicked on close button => passed via `onClose` prop to the `ConfirmationDialog` component.
2. If in the `useEffect` hook it is detected the `state === 'submitting'`  (happens for confirm/cancel click)

Reply:
`handleConfirm` will definitely will called twice but `handleClose` will be called once because the state had already be `submitting` when confirm button was click for the first time.

Reply:
Ah nevermind, for a moment I thought dialog `onClose` would also be called in confirm/cancel, but that's not the case and we explicitly call `handleClose` in those cases. So your reasoning makes perfect sense.

That said, I would still go with the reject/resolve check instead to avoid the additional useEffect and the intermediate submitting state :)

Reply:
I agree that the reject/resolve check is the minimal way to fix it but it felt like bandaging the code so I did not go ahead in that direction.
But, you want me to fix it that way, I can certainly do that.

Reply:
It's surely subjective, but keeping less state feels like less complexity, so I'd appreciate the other solution yeah :)

Reply:
Hey @jonatanklosko , I've pushed the updated code. Please review